{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63273a61-54be-42ad-89f5-f908974a5685",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.decomposition import PCA\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from mpl_toolkits.axes_grid1.axes_divider import make_axes_locatable\n",
    "import networkx as nx\n",
    "from importlib import reload\n",
    "import scipy\n",
    "from sklearn import preprocessing\n",
    "from sklearn import metrics\n",
    "\n",
    "# local imports \n",
    "import graph as gr\n",
    "import utils as ut"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36d2736c-b85e-45e9-838e-8b13ffdfb6ac",
   "metadata": {},
   "source": [
    "# Genome Dynamics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f07a1b6f-eb69-40d3-8fc4-0239fe992289",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputDir =  \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/graph_data/single_cell_expression/\"\n",
    "\n",
    "keys = [\"ND\", \"HFD8\", 'HFD14']\n",
    "macTypes = ['Mac1','Mac2','Mac3','Mac4','Mac5',] # macrophage subtypes\n",
    "\n",
    "data = {}\n",
    "\n",
    "for key in keys:\n",
    "    labelFile  = f\"{inputDir}{key}_labels.pq\"\n",
    "    rnaFile  = f\"{inputDir}{key}_cpm.pq\"\n",
    "\n",
    "    # load the files\n",
    "    lf = pd.read_parquet(labelFile)\n",
    "    rf = pd.read_parquet(rnaFile)\n",
    "\n",
    "    # subset the genes\n",
    "    genes = rf.columns\n",
    "\n",
    "    df = pd.merge(rf, lf, \n",
    "             how='left',\n",
    "             left_index=True,\n",
    "             right_on='cellId')\n",
    "\n",
    "    df['metaType'] = np.where(df['cellType'].isin(macTypes), 'Macrophages', df['cellType'])\n",
    "\n",
    "    X = df[genes].fillna(0)\n",
    "    print(f\"{key} n cells: {X.shape[0]} n genes: {X.shape[1]}\")\n",
    "\n",
    "    data[key] = {\n",
    "        'labels' : lf,\n",
    "        'df' : df,\n",
    "        'X' : X,\n",
    "    }\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5f4d50d-cded-4abf-a7da-e76636af6e6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50d6e83b-4dba-4194-b6bd-24171ccd62e3",
   "metadata": {},
   "source": [
    "# Reduce Dimension"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f24698b4-ad38-47f1-bc97-59b712da675b",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_components = 30 # 90% explained variance\n",
    "# n_components = 0.90 # 90% explained variance\n",
    "pca_args = {\n",
    "    'svd_solver' : 'full',\n",
    "}\n",
    "\n",
    "for key in keys:\n",
    "    X = data[key]['X']\n",
    "    df = data[key]['df']\n",
    "    \n",
    "    # construct the cell type correlation matrices from low-dimensional embeddings\n",
    "    embedding, reducer = gr.reduce_dim(X, n_components, method='pca', **pca_args)\n",
    "\n",
    "    # scale embedding so that it's always positive\n",
    "    embedding = preprocessing.minmax_scale(embedding, feature_range=(0, 1))\n",
    "\n",
    "    embedding = pd.DataFrame(embedding)\n",
    "    features = embedding.columns # get the ebedding column names\n",
    "    \n",
    "    embedding['cellType'] = df['cellType']\n",
    "    embedding['metaType'] = df['metaType']\n",
    "\n",
    "    data[key]['embedding'] = embedding\n",
    "    data[key]['reducer'] = reducer\n",
    "    data[key]['features'] = features\n",
    "\n",
    "    exVar = reducer.explained_variance_ratio_.cumsum()[-1]\n",
    "    print(f\"{key} explained variance: {exVar:.3f} with {n_components=}\")\n",
    "\n",
    "    break\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e42d2271-ae71-425d-bab8-beec0c5d69ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 4, 3\n",
    "\n",
    "colors = {\n",
    "    'ND' : 'darkgreen',\n",
    "    'HFD8' : 'goldenrod',\n",
    "    'HFD14' : 'red',\n",
    "}\n",
    "\n",
    "for key in keys:\n",
    "    reducer = data[key]['reducer']\n",
    "    \n",
    "    sns.lineplot(data=reducer.explained_variance_ratio_.cumsum(),\n",
    "                 marker=\".\",\n",
    "                 mec='k',\n",
    "                 lw=0.5,\n",
    "                 c='k',\n",
    "                 mfc=colors[key])\n",
    "\n",
    "    break\n",
    "\n",
    "sns.despine()\n",
    "plt.ylabel(\"Explained Variance\")\n",
    "plt.xlabel(\"Principal Component\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddf81299-7817-4f07-bf7e-b2818e190c98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae934caa-acdb-4b3c-a79b-8e672430fb09",
   "metadata": {},
   "source": [
    "# correlation matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3562a12f-5847-4d79-bce6-fbb535e6b960",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in keys:\n",
    "    features = data[key]['features']\n",
    "    embedding = data[key]['embedding']\n",
    "\n",
    "    genome = {}\n",
    "\n",
    "    for celltype, group in embedding.groupby('metaType'):\n",
    "        A = group[features].corr()\n",
    "        genome[celltype] = A\n",
    "    \n",
    "        eA, _ = np.linalg.eig(A)\n",
    "        print(f\"{key} {celltype} max lambda: {eA.max():.3f}\")\n",
    "\n",
    "    data[key]['genome'] = genome\n",
    "    print()\n",
    "\n",
    "    break\n",
    "    \n",
    "    # fpath = f\"{savePath}{celltype.replace(' ','_')}_{method}_{n_components}.pq\"\n",
    "    # A.to_parquet(fpath)\n",
    "    # print(f\"{celltype} {A.shape=} at: {fpath}\")\n",
    "    # corrs[celltype] = A\n",
    "    # break\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "952a3fd2-1ded-4b01-be73-483e5963d2e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data['ND'].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2e255cd-72fc-45a3-9e86-23617823f50d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"SAVE THE DATA\"\"\"\n",
    "\n",
    "savePath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/graph_data/correlation_matrices/\"\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    for j, (celltype, A) in enumerate(data[key]['genome'].items()):\n",
    "        print(key, celltype, A.shape)\n",
    "        fpath = f\"{savePath}{celltype.replace(' ','_')}_PCA_{n_components}.pq\"\n",
    "        A.to_parquet(fpath)\n",
    "    break\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37cd8d23-7df6-4c46-990d-d82bcf60271f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b862616a-d5c7-47b9-8856-2509f353308a",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 17, 7.5\n",
    "\n",
    "lut = 5\n",
    "cmap = plt.cm.get_cmap('PuOr', lut=lut)\n",
    "cmap.set_bad(color='lightgrey')\n",
    "\n",
    "fig, axs = plt.subplots(3, 7)\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    for j, (celltype, A) in enumerate(data[key]['genome'].items()):\n",
    "        axs[i, j].imshow(A, cmap=cmap, vmin=-1, vmax=1)\n",
    "        axs[i, j].set_xticks([])\n",
    "        axs[i, j].set_yticks([])\n",
    "        \n",
    "        if i == 0:\n",
    "            axs[i, j].set_title(celltype, fontsize=15)\n",
    "        # if j == 0:\n",
    "            # axs[i, j].set_ylabel(r\"$t0$\", fontsize=15, rotation=0, labelpad=10, ha='right')\n",
    "\n",
    "    break\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be4f78fc-7d87-4361-ac24-e1f0a83cb18b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ut.makeColorbar(cmap, 2, 0.3, 'Correlation', 'horizontal', ['-1', '1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24045ae7-0a35-44af-9d1d-1b9e92dee95f",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = 'ND'\n",
    "ct = 'B cells'\n",
    "\n",
    "A = data[key]['genome'][ct]\n",
    "embedding = data[key]['embedding'] # required for the mean computations\n",
    "features = data[key]['features'] # required for the mean computations\n",
    "\n",
    "edf = embedding[embedding['metaType'] == ct]\n",
    "c = edf[features].mean(axis=0)\n",
    "\n",
    "# # get eignvalues\n",
    "# eA, evA = np.linalg.eigh(A)\n",
    "# print(eA.max())\n",
    "\n",
    "# get means\n",
    "embedding.head()\n",
    "\n",
    "x0, rnorm = scipy.optimize.nnls(A, c, maxiter=100)\n",
    "print(f\"{rnorm=:.3f} {x0.shape=}\")\n",
    "print()\n",
    "\n",
    "dxdt = np.dot(A, x0) - c\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 4, 2\n",
    "plt.plot(dxdt, marker=\".\")\n",
    "plt.axhline(y=0, c='r')\n",
    "# plt.plot(-1*c, marker=\".\")\n",
    "# print(dxdt)\n",
    "# print()\n",
    "# print(c)\n",
    "\n",
    "\n",
    "# plt.rcParams['figure.dpi'] = 300\n",
    "# plt.rcParams['figure.figsize'] = 2, 2\n",
    "\n",
    "# plt.imshow(A)\n",
    "# ax = plt.gca()\n",
    "# ax.axis(False)\n",
    "\n",
    "# divider = make_axes_locatable(ax)\n",
    "# ax2 = divider.append_axes(\"right\", size=\"10%\", pad=\"2%\")\n",
    "\n",
    "# ax2.imshow(x0.reshape(-1, 1))\n",
    "# ax2.axis(False)\n",
    "\n",
    "# # divider = make_axes_locatable(ax)\n",
    "# ax3 = divider.append_axes(\"right\", size=\"10%\", pad=\"20%\")\n",
    "# ax3.imshow(c.to_numpy().reshape(-1, 1))\n",
    "# ax3.axis(False)\n",
    "\n",
    "# eA, _ = np.linalg.eig(A)\n",
    "# eA = np.flip(eA)\n",
    "# eig = pd.DataFrame(eA, columns=['v'])\n",
    "# eig = eig.reset_index(drop=False)\n",
    "\n",
    "# sns.scatterplot(data=eig, \n",
    "#                 x='v',\n",
    "#                 y='index',\n",
    "#                 ax=ax2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0483a133-42d4-49d1-a3bd-d72b3ac918d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e89bb89d-0a89-4c20-bce1-2cb839a673d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 2, 2\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    res = []\n",
    "    for j, (celltype, A) in enumerate(data[key]['genome'].items()):\n",
    "        eA, _ = np.linalg.eig(A)\n",
    "        row = {\n",
    "            'cellType' : celltype,\n",
    "            'maxE' : eA.max()\n",
    "        }\n",
    "        res.append(row)\n",
    "\n",
    "    res = pd.DataFrame(res)\n",
    "    sns.barplot(data=res, \n",
    "                x='cellType',\n",
    "                y='maxE',\n",
    "                palette='viridis',\n",
    "                edgecolor='k')\n",
    "    \n",
    "    plt.gca().tick_params(axis='x', rotation=90)\n",
    "    plt.xlabel(\"\")\n",
    "    plt.ylabel(r\"$\\lambda_{max}$\")\n",
    "    # plt.title('Healthy')\n",
    "    sns.despine()\n",
    "    plt.show()\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2abba6b-6919-438a-9967-b5180eb59170",
   "metadata": {},
   "outputs": [],
   "source": [
    "# celltype = \"Macrophages\"\n",
    "# A = corrs[celltype]\n",
    "# # A = A * -1/\n",
    "# eV, eW = np.linalg.eig(A)\n",
    "# print(f\"for A: {eV.max()}\")\n",
    "\n",
    "# n_cells = 5\n",
    "# # print(f\"{A.shape=}\")\n",
    "\n",
    "# B = np.kron(np.eye(n_cells), A)\n",
    "\n",
    "# eV, eW = np.linalg.eig(B)\n",
    "# print(f\"for B: {eV.max()}\")\n",
    "\n",
    "# B = np.where(B == 0, np.nan, B)\n",
    "\n",
    "\n",
    "# lut = 5\n",
    "# cmap = plt.cm.get_cmap('magma', lut=lut)\n",
    "# cmap.set_bad(color='whitesmoke')\n",
    "\n",
    "# plt.rcParams['figure.dpi'] = 300\n",
    "# plt.rcParams['figure.figsize'] = 5, 5\n",
    "\n",
    "# plt.imshow(B, cmap=cmap, vmin=-1, vmax=1)\n",
    "\n",
    "# plt.title(f\"n={n_cells}\")\n",
    "# _ = plt.yticks([], [])\n",
    "# _ = plt.xticks([], [])\n",
    "# plt.gca().set_aspect('equal')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40c46186-820e-46f1-9386-c4d438136984",
   "metadata": {},
   "source": [
    "# Interaction Dynamics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686e0d94-6f0e-45a3-92ca-24f4422c63b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "dirpath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/graph_data/\"\n",
    "\n",
    "for key in keys:\n",
    "    distpath = f\"{dirpath}/distances/{key}_euclidean_distances.pq\"\n",
    "    labelpath = f\"{dirpath}/global_card_outputs.pq\"\n",
    "    expressonPath =  f\"{dirpath}/spatial_expression/{key}_spatial_cpm.pq\"\n",
    "\n",
    "    # physical distances\n",
    "    D = pd.read_parquet(distpath)\n",
    "    D = D.drop(columns='key')\n",
    "\n",
    "    # spatial data\n",
    "    sdf = pd.read_parquet(expressonPath)\n",
    "\n",
    "    # the cell tye assignments and the coordinates\n",
    "    cdf = pd.read_parquet(labelpath)\n",
    "    cdf = cdf[cdf['key'] == key].reset_index(drop=True)\n",
    "    cdf['Macrophages'] = cdf[macTypes].sum(axis=1)\n",
    "\n",
    "    data[key]['D'] = D\n",
    "    data[key]['sdf'] = sdf\n",
    "    data[key]['cdf'] = cdf\n",
    "\n",
    "    print(f\"{key} {D.shape=} {sdf.shape=} {cdf.shape=}\")\n",
    "\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f11af7f6-0894-4dc2-a412-0149095e022e",
   "metadata": {},
   "source": [
    "# visualization exmaple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "543b121d-8a17-4d3b-9c98-d619b004d278",
   "metadata": {},
   "outputs": [],
   "source": [
    "q = 0.75 # quantile thresholdholding for card output, above this value is a positive hit\n",
    "n_nodes = 100 # number of nodes per sample after thresholding\n",
    "celltype = 'Macrophages'\n",
    "key = 'ND'\n",
    "\n",
    "cdf = data[key]['cdf']\n",
    "sdf = data[key]['sdf']\n",
    "D = data[key]['D']\n",
    "\n",
    "threshold = np.quantile(cdf[celltype], q)\n",
    "print(f\"{key} {celltype} threshold is: {threshold:.4f}\")\n",
    "\n",
    "nbrhd = gr.get_neighborhood(cdf, \n",
    "                            center=True, \n",
    "                            n=n_nodes, \n",
    "                            metric='minkowski')\n",
    "\n",
    "# subset the edges and coords of the spaitial data\n",
    "coords = cdf[cdf['nodeId'].isin(nbrhd)].reset_index()\n",
    "\n",
    "# subset the high-scoring locations for a specific cell type\n",
    "coords['flag'] = np.where(coords[celltype] > threshold, 1, 0)\n",
    "nodeSet = coords['nodeId'].to_list()    \n",
    "\n",
    "# get the edges between the nodes selected above \n",
    "edges = D[(D['node1'].isin(nodeSet)) & (D['node2'].isin(nodeSet))].reset_index()\n",
    "\n",
    "# get gene expression for these nodes\n",
    "stx = sdf[sdf.index.isin(nodeSet)]\n",
    "print(f\"{coords.shape=} {edges.shape=} {stx.shape=}\")\n",
    "\n",
    "corr = []\n",
    "pvals = []\n",
    "flag1 = []\n",
    "flag2 = []\n",
    "\n",
    "# # compute correlations\n",
    "for n1, n2 in edges[['node1', 'node2']].values:\n",
    "    g1 = stx.loc[n1, :].values\n",
    "    g2 = stx.loc[n2, :].values\n",
    "\n",
    "    score, pval = scipy.stats.pearsonr(g1, g2)\n",
    "    corr.append(score)\n",
    "    pvals.append(pval)\n",
    "\n",
    "    f1 = coords.loc[coords['nodeId'] == n1, 'flag'].values[0]\n",
    "    f2 = coords.loc[coords['nodeId'] == n2, 'flag'].values[0]\n",
    "\n",
    "    flag1.append(f1)\n",
    "    flag2.append(f2)\n",
    "\n",
    "\n",
    "\n",
    "edges['flag1'] = flag1\n",
    "edges['flag2'] = flag2\n",
    "edges['flag'] = edges['flag1'] + edges['flag2']\n",
    "print(edges['flag'].value_counts())\n",
    "\n",
    "edges['correlation'] = corr\n",
    "edges['pvals'] = pvals\n",
    "edges['w'] = edges['correlation'] / (edges['d'] ** 0.25)\n",
    "# edges['wt'] = np.where(edges['w'] > 0.15, 1, 0)\n",
    "\n",
    "\n",
    "edges.head()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ab17589-9e00-410e-8326-8de05da4cd8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "G = gr.build_graph(edges, coords)\n",
    "\n",
    "node_attr = coords.set_index('nodeId').to_dict('index')\n",
    "nx.set_node_attributes(G, node_attr)    \n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "# plt.rcParams['figure.facecolor'] = \"none\"\n",
    "plt.rcParams['figure.figsize'] = 8, 4\n",
    "\n",
    "# plotting params\n",
    "node_size = 60\n",
    "lut = 5\n",
    "# cmap = plt.cm.get_cmap('Greys', lut=lut)\n",
    "cmap = plt.cm.get_cmap('RdYlGn', lut=lut)\n",
    "fig, ax = plt.subplots(1, 3)\n",
    "\n",
    "\n",
    "\"\"\"DRAW NODE SELECTION\"\"\"\\\n",
    "# map nodes to colors\n",
    "colorMap = {\n",
    "    1 : 'r',\n",
    "    0 : 'lightgrey',\n",
    "}\n",
    "node_colors = [colorMap[n[1]['flag']] for n in G.nodes(data=True)]\n",
    "node_colors\n",
    "\n",
    "# network plots\n",
    "nx.draw_networkx_nodes(G,\n",
    "                       pos=G.pos,\n",
    "                       node_size=node_size,\n",
    "                       node_color=node_colors,\n",
    "                       edgecolors='k',\n",
    "                       linewidths=1.5,\n",
    "                       ax=ax[0])\n",
    "\n",
    "ax[0].set_aspect('equal')\n",
    "ax[0].axis(False)\n",
    "\n",
    "\"\"\"DISTANCE WEIGHTED GRAPH\"\"\"\n",
    "node_list = [n[0] for n in G.nodes(data=True) if n[1]['flag'] == 1]\n",
    "edge_list = [(e[0], e[1]) for e in G.edges(data=True) if e[2]['flag'] == 2]\n",
    "eweights = [e[2]['w'] for e in G.edges(data=True) if e[2]['flag'] == 2]\n",
    "eweights = preprocessing.minmax_scale(eweights, feature_range=(0, 1))\n",
    "\n",
    "nx.draw_networkx_nodes(G,\n",
    "                       pos=G.pos,\n",
    "                       node_size=node_size,\n",
    "                       node_color=colorMap[1],\n",
    "                       nodelist=node_list,\n",
    "                       edgecolors='k',\n",
    "                       linewidths=1.5,\n",
    "                       ax=ax[1])\n",
    "\n",
    "nx.draw_networkx_edges(G,\n",
    "                       pos=G.pos,\n",
    "                       nodelist=node_list,\n",
    "                       edgelist=edge_list,\n",
    "                       width=eweights*2,\n",
    "                       # edge_color='k',\n",
    "                       edge_color=eweights*5,\n",
    "                       edge_cmap=cmap, \n",
    "                       # alpha=0.75,\n",
    "                       alpha=eweights,\n",
    "                       ax=ax[1])   \n",
    "\n",
    "ax[1].set_aspect('equal')\n",
    "ax[1].axis(False)\n",
    "\n",
    "\"\"\" Spatial Adjancency \"\"\"\n",
    "node_list = [n[0] for n in G.nodes(data=True) if n[1]['flag'] == 1]\n",
    "edge_list = [(e[0], e[1]) for e in G.edges(data=True) if e[2]['flag'] == 2]\n",
    "eweights = [e[2]['w'] for e in G.edges(data=True) if e[2]['flag'] == 2]\n",
    "eweights = preprocessing.minmax_scale(eweights, feature_range=(0, 1))\n",
    "\n",
    "eweights = np.where(eweights > 0.56, 'k', 'none')\n",
    "\n",
    "nx.draw_networkx_nodes(G,\n",
    "                       pos=G.pos,\n",
    "                       node_size=node_size,\n",
    "                       node_color=colorMap[1],\n",
    "                       nodelist=node_list,\n",
    "                       edgecolors='k',\n",
    "                       linewidths=1.5,\n",
    "                       ax=ax[2])\n",
    "\n",
    "nx.draw_networkx_edges(G,\n",
    "                       pos=G.pos,\n",
    "                       nodelist=node_list,\n",
    "                       edgelist=edge_list,\n",
    "                       width=1,\n",
    "                       edge_color=eweights,\n",
    "                       # alpha=0.5,\n",
    "                       ax=ax[2])   \n",
    "\n",
    "ax[2].set_aspect('equal')\n",
    "ax[2].axis(False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2c79975-8509-4a3f-958f-3b8efe6c7413",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Save the spatial data \"\"\"\n",
    "\n",
    "outdir = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/graph_data/for_amit/\"\n",
    "\n",
    "# get the spatial adjancency matrix\n",
    "W = nx.adjacency_matrix(H, weight='w')\n",
    "W = W.todense()\n",
    "W *= (1.0/W.max()) #\n",
    "W = np.where(W > 0.4, 1, 0)\n",
    "\n",
    "nodes = pd.DataFrame.from_dict(dict(H.nodes(data=True)), orient='index')\n",
    "nodes = nodes.drop(columns=['index'])\n",
    "nodes = nodes.reset_index(drop=False)\n",
    "nodes = nodes.rename(columns={'index' : 'nodeId'})\n",
    "\n",
    "\"\"\" Save the node data \"\"\"\n",
    "cols = [\n",
    "    'nodeId', 'x', 'y', 'Macrophages',\n",
    "]\n",
    "\n",
    "fname = f\"{outdir}node_data.csv\"\n",
    "nodes[cols].to_csv(fname, index=False)\n",
    "node_list = nodes['nodeId'].to_list()\n",
    "print(f\"Saved node data: {fname}\")\n",
    "\n",
    "\"\"\" save the spatial adjancency \"\"\"\n",
    "W = pd.DataFrame(W, index=node_list, columns=node_list)\n",
    "W = W.reset_index(drop=False)\n",
    "W = W.rename(columns={'index' : 'nodeId'})\n",
    "\n",
    "fname = f\"{outdir}A.csv\"\n",
    "W.to_csv(fname, index=False)\n",
    "print(f\"Saved node data: {fname}\")\n",
    "\n",
    "\"\"\" save the edge data \"\"\"\n",
    "\n",
    "fedges = edges[(edges['node1'].isin(node_list)) & (edges['node2'].isin(node_list)) ]\n",
    "cols = [\n",
    "    'node1', \n",
    "    'node2', \n",
    "    'd', \n",
    "    'correlation',\n",
    "]\n",
    "\n",
    "fname = f\"{outdir}edges.csv\"\n",
    "fedges[cols].to_csv(fname, index=False)\n",
    "print(f\"Saved node data: {fname}\")\n",
    "\n",
    "# fedges[cols].head()\n",
    "\n",
    "# W = pd.DataFrame()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2644b077-663c-431c-a0a1-47b9e4d4547b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fedges.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e075b1d9-72b2-4951-a969-759dc3f472bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65957456-8b14-4967-a5f5-250ab5fe81a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "A = data[key]['genome'][celltype]\n",
    "A = np.where(A.abs() > 0.5, 1, 0)\n",
    "\n",
    "H = G.edge_subgraph(edge_list)\n",
    "W = nx.adjacency_matrix(H, weight='w')\n",
    "W = W.todense()\n",
    "W *= (1.0/W.max()) #\n",
    "W = np.where(W > 0.4, 1, 0)\n",
    "\n",
    "K = np.kron(W, A)\n",
    "\n",
    "eK, evK = np.linalg.eigh(K)\n",
    "\n",
    "# rowSum = K.sum(axis=1)\n",
    "# dinv = [1/np.sqrt(x) for x in rowSum]\n",
    "\n",
    "# degree = np.diag(dinv)\n",
    "# L = degree - K\n",
    "# L = np.dot(np.dot(degree, L), degree) #normalized laplacian \n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(1, 3)\n",
    "ax[0].imshow(A, cmap='plasma')\n",
    "ax[1].imshow(W)\n",
    "ax[2].imshow(K)\n",
    "# ax[3].imshow(L)\n",
    "\n",
    "ax[0].axis(False)\n",
    "ax[1].axis(False)\n",
    "ax[2].axis(False)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4619bb58-a308-4816-87a7-07c72db5eb6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(np.flip(eK))\n",
    "# plt.xscale('log')\n",
    "# plt.yscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18f4a91b-48c8-4c37-be06-56e3b6048284",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "406e9d4e-9bcf-4d8e-b6c4-81296312e6c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[key].keys()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3a20653-8662-4862-8cb7-6ce4172607a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "666fdc9d-cccc-485d-86ba-b319c19f830f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f548876-0370-45e3-b4e5-a266bf6d3eea",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_size = 1 # number of samples\n",
    "q = 0.9 # quantile thresholdholding for card output, above this value is a positive hit\n",
    "n_nodes = 30 # number of nodes per sample after thresholding\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for key in keys:\n",
    "    cdf = data[key]['cdf']\n",
    "    df = data[key]['df']\n",
    "\n",
    "    # graphs for each cell type\n",
    "    for celltype in sorted(df['metaType'].unique()):\n",
    "        threshold = np.quantile(cdf[celltype], q)\n",
    "        print(f\"{key} {celltype} threshold is: {threshold:.4f}\")\n",
    "        \n",
    "        coords = cdf.copy()\n",
    "        coords['flag'] = np.where(coords[celltype] > threshold, 1, 0)\n",
    "\n",
    "        break\n",
    "    \n",
    "    print()\n",
    "    break\n",
    "\n",
    "    # cdf = cdf[cdf[celltype] > threshold]\n",
    "    \n",
    "    # print(cdf.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a91001a-f922-43e1-914c-2de5116e544e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbe8d2a5-60ce-405d-8723-a207165dce6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(gr)\n",
    "\n",
    "celltype = \"Macrophages\"\n",
    "sample_size = 1 # number of samples\n",
    "nodes = 100 # number of nodes per sample\n",
    "q = 0.7 # quantile tyresholdholding for card output\n",
    "\n",
    "threshold = np.quantile(cdf[celltype], q)\n",
    "print(f\"{celltype} threshold={threshold:.4f} in {key}\")\n",
    "\n",
    "graphs = []\n",
    "\n",
    "for i in range(sample_size):\n",
    "    # point = cdf[['x', 'y']].sample(1).to_numpy().ravel()\n",
    "\n",
    "    nbrhd = gr.get_neighborhood(cdf, \n",
    "                                center=True, \n",
    "                                n=nodes, \n",
    "                                metric='minkowski')\n",
    "\n",
    "    # subset the edges and coords of the spaitial data\n",
    "    coords = cdf[cdf['nodeId'].isin(nbrhd)].reset_index()\n",
    "\n",
    "    # subset the high-scoring locations for a specific cell type\n",
    "    # coords = coords[coords[celltype] > threshold]\n",
    "    nodeSet = coords['nodeId'].to_list()    \n",
    "\n",
    "    # get the edges between the nodes selected above \n",
    "    edges = D[(D['node1'].isin(nodeSet)) & (D['node2'].isin(nodeSet))].reset_index()\n",
    "\n",
    "\n",
    "    # get gene expression for these nodes\n",
    "    stx = sdf[sdf.index.isin(nodeSet)]\n",
    "    print(f\"{coords.shape=} {edges.shape=} {stx.shape=}\")\n",
    "\n",
    "    corr = []\n",
    "    pvals = []\n",
    "\n",
    "    # # compute correlations\n",
    "    for n1, n2 in edges[['node1', 'node2']].values:\n",
    "\n",
    "        g1 = stx.loc[n1, :].values\n",
    "        g2 = stx.loc[n2, :].values\n",
    "\n",
    "        score, pval = scipy.stats.pearsonr(g1, g2)\n",
    "        corr.append(score)\n",
    "        pvals.append(pval)\n",
    "    \n",
    "\n",
    "    edges['correlation'] = corr\n",
    "    edges['pvals'] = pvals\n",
    "    edges['w'] = edges['correlation'] / (edges['d'] ** 0.25)\n",
    "    edges['wt'] = np.where(edges['w'] > 0.15, 1, 0)\n",
    "    \n",
    "    G = gr.build_graph(edges, coords)\n",
    "    graphs.append(G)\n",
    "\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6e17461-db29-4f9e-9724-3813b7c40615",
   "metadata": {},
   "outputs": [],
   "source": [
    "node_colors = G.nodes()\n",
    "node_colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0647ff5-e408-4655-8eb8-e184c72a9e39",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "G = graphs[0]\n",
    "relationship = 'w'\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.facecolor'] = \"none\"\n",
    "plt.rcParams['figure.figsize'] = 9, 9\n",
    "\n",
    "# plotting params\n",
    "lut = 5\n",
    "# cmap = plt.cm.get_cmap('Greys', lut=lut)\n",
    "cmap = plt.cm.get_cmap('RdYlGn', lut=lut)\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "node_colors = []\n",
    "\n",
    "\n",
    "for n in G.nodes():\n",
    "    nodeRecord = coords[coords['nodeId'] == n]\n",
    "\n",
    "    if nodeRecord[celltype].values[0] > threshold:\n",
    "        node_colors.append(\"r\")\n",
    "    else:\n",
    "        node_colors.append(\"lightgrey\")\n",
    "    \n",
    "    # break\n",
    "\n",
    "\n",
    "# network plots\n",
    "nx.draw_networkx_nodes(G,\n",
    "                       pos=G.pos,\n",
    "                       node_size=150,\n",
    "                       # node_color='lightgrey',\n",
    "                       node_color=node_colors,\n",
    "                       edgecolors='k',\n",
    "                       linewidths=1.5,\n",
    "                       ax=ax)\n",
    "\n",
    "eweights = np.array([e[relationship] for node1, node2, e in G.edges(data=True)])\n",
    "eweights = preprocessing.minmax_scale(eweights, feature_range=(0, 1))\n",
    "\n",
    "\n",
    "# print(eweights)\n",
    "\n",
    "# nx.draw_networkx_edges(G,\n",
    "#                        pos=G.pos,\n",
    "#                        # width=3,\n",
    "#                        width=eweights*6,\n",
    "#                        # edge_color='k',\n",
    "#                        edge_color=eweights,\n",
    "#                        edge_cmap=cmap, \n",
    "#                        # alpha=0.75,\n",
    "#                        alpha=eweights,\n",
    "#                        ax=ax)   \n",
    "\n",
    "# sm = plt.cm.ScalarMappable(cmap=cmap, norm=plt.Normalize(vmin = -1, vmax=1))\n",
    "# sm._A = []\n",
    "# plt.colorbar(sm)\n",
    "\n",
    "ax.set_aspect('equal')\n",
    "ax.axis(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "897a64f7-9b14-4025-be00-4449bd97ce82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# np.isnan(edgecolors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9feabb6-bbe5-4c55-b3f5-ea88a0ff5625",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the adjacency matrix\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.facecolor'] = \"none\"\n",
    "plt.rcParams['figure.figsize'] = 9, 9\n",
    "\n",
    "Adj = nx.adjacency_matrix(G, weight='w')\n",
    "Adj = Adj.todense()\n",
    "Adj *= (1.0/Adj.max()) # need something different for arrays\n",
    "\n",
    "lut = 5\n",
    "cmap = plt.cm.get_cmap('viridis', lut=lut)\n",
    "cmap.set_bad(color='whitesmoke')\n",
    "\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "\n",
    "axs[0].imshow(Adj, cmap, vmin=0, vmax=1)\n",
    "\n",
    "axs[0].set_yticks([], [])\n",
    "axs[0].set_xticks([], [])\n",
    "axs[0].set_aspect('equal')\n",
    "\n",
    "\n",
    "Adj_binary = np.where(Adj > 0.6, 1, 0)\n",
    "axs[1].imshow(Adj_binary, cmap='binary')\n",
    "axs[1].set_yticks([], [])\n",
    "axs[1].set_xticks([], [])\n",
    "axs[1].set_aspect('equal')\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30083c95-eb2e-4ee0-8434-68644277e27a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the adjacency matrix\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.facecolor'] = \"none\"\n",
    "plt.rcParams['figure.figsize'] = 9, 9\n",
    "\n",
    "Adj = nx.adjacency_matrix(G, weight='d')\n",
    "Adj = Adj.todense()\n",
    "Adj *= (1.0/Adj.max()) # need something different for arrays\n",
    "\n",
    "lut = 5\n",
    "cmap = plt.cm.get_cmap('coolwarm', lut=lut)\n",
    "cmap.set_bad(color='whitesmoke')\n",
    "\n",
    "\n",
    "plt.imshow(Adj, cmap, vmin=0, vmax=1)\n",
    "\n",
    "axs = plt.gca()\n",
    "\n",
    "axs.set_yticks([], [])\n",
    "axs.set_xticks([], [])\n",
    "axs.set_aspect('equal')\n",
    "\n",
    "\n",
    "# Adj_binary = np.where(Adj > 0.6, 1, 0)\n",
    "# axs[1].imshow(Adj_binary, cmap='binary')\n",
    "# axs[1].set_yticks([], [])\n",
    "# axs[1].set_xticks([], [])\n",
    "# axs[1].set_aspect('equal')\n",
    "\n",
    "# plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08591bbd-1f68-426d-8d16-5801430feebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ac279ab-2f69-407e-8cf3-28142261f793",
   "metadata": {},
   "outputs": [],
   "source": [
    "ldf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa27f695-fb43-4df6-abb4-61998854551f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afc80f12-a353-49c9-becb-7c18292c51da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab2f9f94-1a8f-4c84-983b-eb948aeaef88",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d401ce30-84a6-487d-affc-9208360ac606",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a63b45b5-b643-4674-8c8b-3c9cc8a9f30f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24fdc1a3-2be2-4c49-a1a3-4b63f5e4f33e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "def3f556-c2b2-404c-9fdb-a7a86d49e397",
   "metadata": {},
   "outputs": [],
   "source": [
    "# n = len(corrs) * n_components\n",
    "# B = np.zeros((n, n))\n",
    "\n",
    "# # plotting stuff\n",
    "# yticks = []\n",
    "# ylabels = []\n",
    "# lines = []\n",
    "\n",
    "# for i, (celltype, A) in enumerate(corrs.items()):\n",
    "#     start = i * n_components\n",
    "#     end = ((i + 1) * n_components)\n",
    "#     midpoint = (end + start) / 2\n",
    "#     yticks.append(midpoint)\n",
    "#     ylabels.append(celltype)\n",
    "#     lines.append(end)\n",
    "\n",
    "#     # build the block matrix\n",
    "#     B[start:end, start:end] = A\n",
    "    \n",
    "\n",
    "# plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.imshow(B, \n",
    "#            cmap='magma',\n",
    "#            vmin=0)    \n",
    "# _ = plt.yticks(yticks, ylabels)\n",
    "# _ = plt.xticks([], [])\n",
    "\n",
    "# for l in lines:\n",
    "#     plt.axvline(x=l-0.5, c='w', lw=1)\n",
    "#     plt.axhline(y=l-0.5, c='w', lw=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61c33d76-981b-46c1-add1-a38516520bb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 15, 3\n",
    "\n",
    "fig, axs = plt.subplots(1, len(corrs))\n",
    "\n",
    "for i, (celltype, A) in enumerate(corrs.items()):\n",
    "    axs[i].imshow(A, cmap='magma', vmin=-1, vmax=1)\n",
    "    axs[i].set_title(celltype)\n",
    "    axs[i].axis(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c431af5-ae4a-4fff-b186-f086327d9914",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype = \"Macrophages\"\n",
    "A = corrs[celltype]\n",
    "n_cells = 5\n",
    "print(f\"{A.shape=}\")\n",
    "\n",
    "B = np.kron(np.eye(n_cells), A)\n",
    "\n",
    "plt.imshow(B, cmap='coolwarm')    \n",
    "\n",
    "_ = plt.yticks([], [])\n",
    "_ = plt.xticks([], [])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b80b9d8-e1b6-4dd2-88c8-51f83caef47a",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = \"ND\"\n",
    "\n",
    "dirpath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/graph_data/\"\n",
    "coordpath = f\"{dirpath}coordinates.pq\"\n",
    "distpath = f\"{dirpath}/distances/{key}_euclidean_distances.pq\"\n",
    "edgepath = f\"{dirpath}/edge_lists/{key}_harmonic_highlevel_edgelist.pq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8184cb62-a4bc-4fb1-8dda-71bdf81c3de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7be89594-0f46-4ca1-9668-805911b6735e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793c95ab-4e00-4174-ad62-bf2d0e250f80",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b2007a0-bccc-487e-ba44-090262ca84cb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd03a3f1-e821-424c-bd47-0c02a5da3947",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # build an arbitrary sizes A matrix for a sinfle cell type\n",
    "\n",
    "# celltype = \"Macrophages\"\n",
    "# A = corrs[celltype]\n",
    "# print(f\"{A.shape=}\")\n",
    "# n_cells = 5\n",
    "# n = n_cells * n_components\n",
    "# B = np.zeros((n, n))\n",
    "# print(f\"{B.shape=}\")\n",
    "\n",
    "# # plotting stuff\n",
    "# yticks = []\n",
    "# ylabels = []\n",
    "# lines = []\n",
    "\n",
    "# for i in range(n_cells):\n",
    "#     start = i * n_components\n",
    "#     end = ((i + 1) * n_components)\n",
    "#     midpoint = (end + start) / 2\n",
    "\n",
    "#     label = f\"{celltype}_{i+1}\"\n",
    "    \n",
    "#     yticks.append(midpoint)\n",
    "#     ylabels.append(label)\n",
    "#     lines.append(end)\n",
    "    \n",
    "#     B[start:end, start:end] = A\n",
    "    \n",
    "# plt.rcParams['figure.dpi'] = 300\n",
    "\n",
    "# plt.imshow(B, \n",
    "#            cmap='magma',)    \n",
    "# _ = plt.yticks(yticks, ylabels)\n",
    "# _ = plt.xticks([], [])\n",
    "\n",
    "# for l in lines:\n",
    "#     plt.axvline(x=l-0.5, c='w', lw=1)\n",
    "#     plt.axhline(y=l-0.5, c='w', lw=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f5adca-9c18-4af9-b441-9e73231c6492",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype = \"Macrophages\"\n",
    "A = corrs[celltype]\n",
    "n_cells = 5\n",
    "print(f\"{A.shape=}\")\n",
    "\n",
    "B = np.kron(np.eye(n_cells), A)\n",
    "\n",
    "plt.imshow(B, cmap='coolwarm')    \n",
    "\n",
    "_ = plt.yticks([], [])\n",
    "_ = plt.xticks([], [])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee1c8cab-ac93-4a94-ab50-936da59b2faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = \"ND\"\n",
    "\n",
    "dirpath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/graph_data/\"\n",
    "coordpath = f\"{dirpath}coordinates.pq\"\n",
    "distpath = f\"{dirpath}/distances/{key}_euclidean_distances.pq\"\n",
    "edgepath = f\"{dirpath}/edge_lists/{key}_harmonic_highlevel_edgelist.pq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba66e5ce-1e75-4f6b-9260-dedc565fb002",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f898125-7c1d-400f-85fa-1c65da3533c3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbd4c47b-3a1f-41ce-b062-d7b644f4b66b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3e80d74-f392-46f6-b2b6-9bd120f494a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efd6f4be-11d9-466c-b66d-78d3ec6ee0ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype = \"Macrophages\"\n",
    "A = corrs[celltype]\n",
    "n_cells = 5\n",
    "B = np.kron(np.eye(n_cells), A)\n",
    "\n",
    "print(f\"{A.shape=}\")\n",
    "print(f\"{B.shape=}\")\n",
    "\n",
    "eA, _ = np.linalg.eigh(A)\n",
    "eB, _ = np.linalg.eigh(B)\n",
    "\n",
    "print(eA.round(2))\n",
    "\n",
    "# plt.rcParams['figure.dpi'] = 300\n",
    "# plt.rcParams['figure.figsize'] = 3, 3\n",
    "# sns.lineplot(data=np.flip(eA),\n",
    "#              marker=\".\", \n",
    "#              markeredgecolor='k',\n",
    "#              lw=1)\n",
    "\n",
    "# sns.lineplot(data=np.flip(eB),\n",
    "#              marker=\".\", \n",
    "#              markeredgecolor='k',\n",
    "#              lw=1,\n",
    "#              color='C1')\n",
    "\n",
    "# sns.despine()\n",
    "# # plt.xscale('log')\n",
    "# plt.yscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fe36aa6-fb40-40eb-b5d2-9c59d96a1862",
   "metadata": {},
   "outputs": [],
   "source": [
    "eA, _ = np.linalg.eig(-1*A)\n",
    "eA.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aac293d1-5f9e-4907-9b9e-172158292c0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "eA, _ = np.linalg.eig(A)\n",
    "eA.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6399c73-a8d5-416b-858d-baaa2ad17d61",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
